docs/design/direct-assign-volume.md

```diff
@@ -0,0 +1,139 @@
+will make sure backing storage is attached to the node (often a block device), and then a CSI node driver will
+create the filesystem (if needed) and mount it.
+
+![Volume handling without direct-assignment](arch-images/DA-not-enabled.png)
+
+Ideally, for a VM, we will consume the block device itself and not utilize the host mount. This simplifies volume management
+on the host, and is more efficient from a storage IO perspective. 
+
+![Volume handling with direct-assignment](arch-images/DA-UC.png)
+
+# Proposal - Introduce pattern for directly assigning block device for mounting in the guest
+
+In the existing upstream, there is no direct communication between CSI and the container runtime. By the
+time volume information is presented to the runtime, it is a path on the given host filesystem. We need a method to:
+ - enlighten the CSI driver that a given PVC is to be utilized by Kata (or any other runtime that would prefer to mount the volume on their own).
+ - communication path between the CSI node driver and the runtime to detail: what the backing block device is,
+   what kind of filesystem it is, and where it should be mounted with respect to the container rootfs.
+
+Ideally we'd be able to come up with a solution that would facilitate a runtime calling into CSI instead of Kubelet, as is done today for CNI. Working within the existing constraints that exist, we are proposing a pattern that can be used to facilitate direct-volume assignment, while working with other upstream projects for a more direct method. 
+
+## Carrying out the mount
+
+The CSI node driver will need to be enlightened that the volume will be consumed by a VM-based runtime, and will need 
+to communicate with the runtime to describe where the block device lives, what filesystem is on it, and where it
+should be mounted within the container(s).
+
+### Specifiying that the volume will be consumed by a VM-friendly runtime
+
+If a persistent volume [(PV)](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) is intended to be consumed by a VM-based runtime, the user can specify this by adding an annotation to their respect persistent volume claim (PVC). Example:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    skip-hostmount: "true"
+  name: amazing-kata-volume
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: 4Gi
+```
+
+Based on this annotation, the CSI node driver would identify that the volume should not be mounted, and special handling is needed.
+
+### Communicating block device details to the runtime
+
+The runtime will need to identify that a mount is backed by a block device and that it needs to handle the direct-assignment use case.
+
+When a volume is presented for direct-assignment, the CSI driver will create a particular file, `csiPlugin.json`
+at the root of the volume on the host. The contents of this file will be defined as follows:
```

Comment:
As pointed out already, the potential risk here is that a user uses a normal CSI driver and mounts a volume which contains this special file to get access to devices.

I'd like to add that checking for "is the volume mounted" is not sufficient to prevent this. The CSI driver could be one which unpacks an image or tar archive at the volume path, without mounting.

What the runtime needs some additional meta data about the volume before looking into its content.

Reply:
I see, thanks for clarifying the alternative CSI driver scenario. I'd definitely consider this a limitation with the quick proposal, and something to be addressed by longer term proposal (ie, path for additional metadata to be provided to runtime)

Reply:
I think with this gap, we may want to table this proposal for the short term.